src/init/AppRunner.ts

```diff
@@ -54,99 +80,81 @@ export class AppRunner {
           process.exit(1);
-        },
-      ).catch((error): void => {
-        this.logger.error(`Could not start server: ${error}`, { error });
-        process.exit(1);
-      });
+        }))
+      .catch((error): void => this.writeError(stderr, `Could not build the config files from ${config}`, error));
   }
 
   /**
-   * Creates the main app object to start the server from a given config.
-   * @param loaderProperties - Components.js loader properties.
-   * @param configFile - Path to a Components.js config file.
-   * @param variables - Variables to pass into the config file.
+   * Writes the given message and error to the `stderr` and exits the process.
    */
-  public async createApp(
-    loaderProperties: IComponentsManagerBuilderOptions<App>,
-    configFile: string,
-    variables: CliParams | Record<string, any>,
-  ): Promise<App> {
-    // Translate command-line parameters if needed
-    if (typeof variables.loggingLevel === 'string') {
-      variables = this.createVariables(variables as CliParams);
-    }
+  private writeError(stderr: WriteStream, message: string, error: Error): never {
+    stderr.write(`${message}\nCause:\n${createErrorMessage(error)}\n`);
+    stderr.write(`${error.stack}\n`);
+    process.exit(1);
+  }
 
-    // Set up Components.js
-    const componentsManager = await ComponentsManager.build(loaderProperties);
-    await componentsManager.configRegistry.register(configFile);
+  /**
+   * Handles the first Components.js instantiation,
+   * where CLI settings and variable mappings are created.
+   */
+  private async resolveVariables(componentsManager: ComponentsManager<CliResolver>, argv: string[]):
+  Promise<Record<string, unknown>> {
+    // Create a resolver, which combines a CliExtractor and a VariableResolver
+    const resolver = await componentsManager.instantiate(DEFAULT_CLI_RESOLVER, {});
+    // Extract values from CLI parameters
+    if (!resolver.extractor) {
+      throw new Error('No CliExtractor is defined.');
+    }
+    const cliValues = await resolver.extractor.handleSafe(argv);
+    //  Convert CLI values into variable mappings
+    return await resolver.resolver.handleSafe(cliValues);
+  }
 
-    // Create the app
-    const app = 'urn:solid-server:default:App';
-    return await componentsManager.instantiate(app, { variables });
+  /**
+   * The second Components.js instantiation,
+   * where the App is created and started using the variable mappings.
+   */
+  private async startApp(componentsManager: ComponentsManager<App>, variables: Record<string, unknown>): Promise<void> {
+    // Create app
+    const app = await componentsManager.instantiate(DEFAULT_APP, { variables });
+    // Execute app
+    await app.start();
   }
 
   /**
-   * Translates command-line parameters into Components.js variables.
+   * Creates the Components Manager that will be used for instantiating.
+   * Typing is set to `any` since it will be used for instantiating multiple objects.
```

Comment:
It can, but in practice `T` is then going to be set to `any` though since we use it to build both a `CliResolver` and an `App`. And `<CliResolver | App>` would make the code much more ugly since we would need a lot more checks.

Reply:
Actually it does make sense to use `<T>` here since it's a public function.

Reply:
Yeah, and then we can just call it as `createComponentsManager<MyThing>(…)`, right?